packages/server-wallet/src/wallet/index.ts

```diff
@@ -165,34 +158,70 @@ export class Wallet implements WalletInterface {
       throw err;
     }
 
-    const {outbox} = await protocolEngine(channelIds);
-
-    return {
-      channelResults: [
-        {
-          outbox,
-          appData: '',
-          appDefinition: '',
-          channelId: '',
-          challengeExpirationTime: 0,
-          status: 'funding',
-          participants: [],
-          allocations: [],
-          turnNum: 0,
-        },
-      ],
-    };
+    const {channelResults, outbox} = await takeActions(channelIds);
+
+    return {outbox, channelResults};
   }
   onNotification(_cb: (notice: Notification) => void): {unsubscribe: () => void} {
     throw 'Unimplemented';
   }
 }
 
-type ExecutionResult = {ids: Bytes32[]; outbox: Outgoing[]};
+type ExecutionResult = {
+  outbox: Outgoing[];
+  channelResults: ClientChannelResult[];
+  error?: any;
+};
+
+const takeActions = async (channels: Bytes32[]): Promise<ExecutionResult> => {
+  const outbox: Outgoing[] = [];
+  const channelResults: ClientChannelResult[] = [];
+  let error: Error | undefined = undefined;
+  while (channels.length && !error) {
+    const tx = await knex.transaction();
+
+    const setError = async (e: Error): Promise<void> => {
+      error = e;
+      await tx.rollback();
+    };
+    const markChannelAsDone = async (): Promise<any> => {
+      channels.shift() as string;
+    };
+
+    const handleAction = async (action: ProtocolAction): Promise<any> => {
+      switch (action.type) {
+        case 'SignState': {
+          const notices = await Store.signState(action, tx);
+          notices.map(n => outbox.push(n.notice));
+
+          const {channelResult} = await Channel.forId(action.channelId, tx);
+          channelResults.push(channelResult);
```

Comment:
This just generally feel unintuitive. Maybe I'm thinking about it wrong, but it seems like there should be a separation of interfaces between the Wallet, the Store, and the Channel model. Like
```
Wallet (update store, decides what messages / notifications should be)
|__> Store (makes DB-specific optimizations and interacts with model e.g., transactions)
       |__> Channel (implements object in DB-specific domain e.g., PostgreSQL)
```
It seems here we have
```
Wallet (update store, re-uses what store tells it to use as notifications, queries Models directly for response)
|__> Store (makes DB-specific optimizations and interacts with model, relays notifications from model)
       |__> Channel (implements object in DB-specific domain and compiles notifications to be sent)
|__> Channel
```

Reply:
Yup. The store was introduced near the end of this piece of work. I believe that the `Channel` model should live entirely behind a store interface. This decouples the wallet from the database.